pkg/kv/kvserver/raft_apply_asserter.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2019 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// TODO: put in a test package
+package kvserver
+
+import (
+	fmt "fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"go.etcd.io/raft/v3/raftpb"
+)
+
+// RaftApplyAsserter is a test utility that tracks application of Raft
+// commands, and asserts that a command is never applied at different log
+// indexes (i.e. a double-apply or replay).
+type RaftApplyAsserter struct {
+	mu syncutil.Mutex
+	// appliedCmds is a map of applied command IDs to the range ID and Raft log
+	// index at which it was applied.
+	appliedCmds map[kvserverbase.CmdIDKey]*appliedCmd
+}
+
+type appliedCmd struct {
+	cmdID    kvserverbase.CmdIDKey
+	rangeID  roachpb.RangeID
+	logIndex kvpb.RaftIndex
+	replicas map[roachpb.ReplicaID]bool
+}
+
+func NewRaftApplyAsserter() *RaftApplyAsserter {
+	return &RaftApplyAsserter{
+		appliedCmds: map[kvserverbase.CmdIDKey]*appliedCmd{},
+	}
+}
+
+// Applied handles a command being applied, recording it and asserting that it
+// was not already applied at a different log index.
+//
+// TODO: we need to handle snapshots too, by treating all commands up to the
+// snapshot index as applied on that replica.
+func (a *RaftApplyAsserter) Applied(
```

Comment:
nit: consider ordering the coordinate parameters from big to small, i.e. (cmdID, rangeID->replicaID->index)